x/sync/README.md

```diff
@@ -1,3 +1,124 @@
+If it's valid, the key-value pairs in the proof are written to the database.
+If it's not, the client drops the proof and requests the proof from another server. 
+
+A range proof sent by a server must return a continuous range of the key-value pairs, but may not 
+return the full range that was requested.
+For example, a client might request all the key-value pairs in [`requested_start`, `requested_end`] 
+but only receive those in range [`requested_start`, `proof_end`] where `proof_end` < `requested_end`.
+There might be too many key-value pairs to include in one message, or the server may be too busy to provide any more in its response.
+Unless the database is very small, this means that the range proof the client receives in response to
+ its range proof request for the entire database will not contain all of the key-value pairs in the database.
+
+If a client requests a range proof for range [`requested_start`, `requested_end`] but only receives 
+a range proof for [`requested_start`, `proof_end`] where `proof_end` < `requested_end`
+it recognizes that it must still fetch all of the keys in [`proof_end`, `requested_end`]. 
+It repeatedly requests range proofs for chunks of the remaining key range until it has all of the 
+key-value pairs in [`requested_start`, `requested_end`].
+Note that the client may split the remaining key range into chunks and fetch chunks of key-value pairs in parallel.
+
+Additional commits to the database may occur while the client is syncing.
+The sync client can be notified that the root hash of the database it's trying to sync to has changed.
+Detecting that the root hash to sync to has changed is done outside this package.
+For example, if the database is being used to store blockchain state then the sync client would be 
+notified when a new block is accepted because that implies a commit to the database.
+If this occurs, the key-value pairs the client has learned about via range proofs may no longer be up to date.
+
+We use change proofs to correct the out of date key-value pairs. When the sync client is notified that
+the root hash to sync to has changed, it requests a change proof from a server for a given key range.
+For example, if a client has the key-value pairs in range [`start`, `end`] that were in the database 
+when it had `root_hash`, then it will request a change proof that provides all of the key-value changes 
+in range [`start`, `end`] from the database version with root hash `root_hash` to the database version with root hash `new_root_hash`.
+The client verifies the change proof, and if it's valid, it applies the changes to its database.
+If it's not, the client drops the proof and requests the proof from another server.
+
+A server needs to have history in order to serve a change proof.
+Namely, it needs to know all of the database changes between two roots.
+If the server does not have sufficient history to generate a change proof, it will send a range proof instead.
```

Comment:
```suggestion
If the server does not have sufficient history to generate a change proof, it will send a range proof of the `new_root_hash` instead.
```